### PR TITLE
Refactor source db access.

### PR DESCRIPTION
**Pull request by amit-yb**

### CODE REVIEW

File: yb_migrate/src/srcdb/source.go [hunk 1]
```diff
@@ -31,6 +37,26 @@ type Source struct {
...
+
+	sourceDB SourceDB
+}
+
+func (s *Source) DB() SourceDB {
+	if s.sourceDB != nil {
+		return s.sourceDB
+	}
+
+	s.Lock()
```
Comment (sanyamsinghal) [inline yb_migrate/src/srcdb/source.go:51]: why Lock() is needed here?

File: yb_migrate/src/srcdb/source.go [hunk 2]
```diff
@@ -104,3 +130,108 @@ func (s *Source) ParseURI() {
...
+		return tls.Config{
+			RootCAs:            rootCertPool,
+			Certificates:       clientCert,
+			InsecureSkipVerify: false,
+			ServerName:         source.Host,
+		}
+	}
+}
+
+func generateSSLQueryStringIfNotExists(s *Source) string {
```
Comment (sanyamsinghal) [inline yb_migrate/src/srcdb/source.go:207]: nit: it is postgres specific, can be moved to `srcdb/postgres.go`

File: yb_migrate/src/srcdb/source.go [hunk 3]
```diff
@@ -104,3 +130,108 @@ func (s *Source) ParseURI() {
 
 	}
 }
+
+func parseSSLString(source *Source) {
```
Comment (sanyamsinghal) [inline yb_migrate/src/srcdb/source.go:137]: parseSSLString() and createTLSConf() is for mysql only, can be moved to srcdb/mysql.go

File: yb_migrate/src/srcdb/utils.go [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package srcdb
+
+import (
+	"fmt"
+	"log"
+	"os"
+)
+
+func ErrExit(formatString string, args ...interface{}) {
```
Comment (sanyamsinghal) [inline yb_migrate/src/srcdb/utils.go:10]: is there some issue in using this function from utils package directly?

File: yb_migrate/src/srcdb/mysql.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+	db *sql.DB
+}
+
+func newMySQL(s *Source) *MySQL {
+	return &MySQL{source: s}
+}
+
+func (ms *MySQL) Connect() error {
+	db, err := sql.Open("mysql", ms.getConnectionString())
```
Comment (sanyamsinghal) [inline yb_migrate/src/srcdb/mysql.go:22]: i think we can not using this connection outside the Connect() function.
so its good to have `defer db.Close()`

Comment (sanyamsinghal) [reply yb_migrate/src/srcdb/mysql.go:22]: similarly for other sources

**Review by sanyamsinghal — APPROVED** [review]
LGTM, left some comments please respond/take-care before merge

File: yb_migrate/src/srcdb/mysql.go (hunk 5 above)
Comment (amit-yb) [reply yb_migrate/src/srcdb/mysql.go:22]: That's intentional. We want all source db access to happen behind the `SourceDB` implementation. Our eventual plan is to encapsulate ALL source db access inside the `srcdb` package.

That will make it easy to add support for other databases.

File: yb_migrate/src/srcdb/source.go (hunk 1 above)
Comment (amit-yb) [reply yb_migrate/src/srcdb/source.go:51]: This is a standard implementation of a Singleton pattern. If two go-routines call `source.DB()` at (almost) the same time, there is a possibility that we may instantiate the `SourceDB` twice. That's prevented by the lock.

File: yb_migrate/src/srcdb/source.go (hunk 3 above)
Comment (amit-yb) [reply yb_migrate/src/srcdb/source.go:137]: These functions are only accessing the `Source` type. Hence I kept them closer to the `Source` type. I will consider making them methods of the `Source` type. For now, I will leave them as they are here.

File: yb_migrate/src/srcdb/source.go (hunk 2 above)
Comment (amit-yb) [reply yb_migrate/src/srcdb/source.go:207]: Same as above.

File: yb_migrate/src/srcdb/utils.go (hunk 4 above)
Comment (amit-yb) [reply yb_migrate/src/srcdb/utils.go:10]: As of now, `utils` is importing `srcdb` for the `Source` type. Importing `utils` in the `srcdb` package will create a circular import.

If I manage to remove `srcdb` imports from the `utils` package, I will consider using `utils` in this package.

Comment (sanyamsinghal) [reply yb_migrate/src/srcdb/utils.go:10]: got it

**Pull request merged**